src/C++/DataDictionary.cpp

```diff
@@ -592,8 +592,8 @@ void DataDictionary::addXMLGroup( DOMDocument* pDoc, DOMNode* pNode,
 {
   DOMAttributesPtr attrs = pNode->getAttributes();
   std::string name;
-  if(!attrs->get("name", name))
-    throw ConfigError("No name given to group");
+  attrs->get("name", name);
+
```

Comment:
The group node name check happens before this execution.  There are multiple checks for checking if the name attribute on Fields/Groups xml nodes exists before reaching this section.  If the name attribute on a group node is missing, it will have already received a ConfigError by this point.

Reply:
Can you show where this is and if it is truly guaranteed? It may be true, but it seems I would need to keep track of a relatively complex flow to know this is true. I'm uncomfortable removing a check for pulling data from a data structure that can theoretically not be there just because some other method out of context might have done the same check.

As a method, I also can make no guarantees who will call me and what checks they may or may not have done. If the code ever gets restructured, this create a big problem. I can only know my own context and from my point of view I have no idea if the value is there.

Reply:
Here are some following examples 

Lines 319
``
    while(pHeaderFieldNode.get())
    {
      if(pHeaderFieldNode->getName() == "field" || pHeaderFieldNode->getName() == "group" )
      {
        DOMAttributesPtr attrs = pHeaderFieldNode->getAttributes();
        std::string name;
        if(!attrs->get("name", name))
          throw ConfigError("<field> does not have a name attribute");
        std::string required = "false";
        attrs->get("required", required);
        addHeaderField(lookupXMLFieldNumber(pDoc.get(), name), required == "true");
      }
      if(pHeaderFieldNode->getName() == "group")
      {
        DOMAttributesPtr attrs = pHeaderFieldNode->getAttributes();
        std::string required;
        attrs->get("required", required);
        bool isRequired = (required == "Y" || required == "y");
        addXMLGroup(pDoc.get(), pHeaderFieldNode.get(), "_header_", *this, isRequired);
      }

      RESET_AUTO_PTR(pHeaderFieldNode, pHeaderFieldNode->getNextSiblingNode());
    }
``
Line 354

``
Lines 354
    while(pTrailerFieldNode.get())
    {
      if(pTrailerFieldNode->getName() == "field" || pTrailerFieldNode->getName() == "group" )
      {
        DOMAttributesPtr attrs = pTrailerFieldNode->getAttributes();
        std::string name;
        if(!attrs->get("name", name))
          throw ConfigError("<field> does not have a name attribute");
        std::string required = "false";
        attrs->get("required", required);
        addTrailerField(lookupXMLFieldNumber(pDoc.get(), name), required == "true");
      }
      if(pTrailerFieldNode->getName() == "group")
      {
        DOMAttributesPtr attrs = pTrailerFieldNode->getAttributes();
        std::string required;
        attrs->get("required", required);
        bool isRequired = (required == "Y" || required == "y");
        addXMLGroup(pDoc.get(), pTrailerFieldNode.get(), "_trailer_", *this, isRequired);
      }

      RESET_AUTO_PTR(pTrailerFieldNode, pTrailerFieldNode->getNextSiblingNode());
    }
``
Line 402
``
Line 402
      while( pMessageFieldNode.get() )
      {
        if(pMessageFieldNode->getName() == "field"
           || pMessageFieldNode->getName() == "group")
        {
          DOMAttributesPtr attrs = pMessageFieldNode->getAttributes();
          std::string name;
          if(!attrs->get("name", name))
            throw ConfigError("<field> does not have a name attribute");
          int num = lookupXMLFieldNumber(pDoc.get(), name);
          addMsgField(msgtype, num);

          std::string required;
          if(attrs->get("required", required)
             && (required == "Y" || required == "y"))
          {
            addRequiredField(msgtype, num);
          }
        }
        else if(pMessageFieldNode->getName() == "component")
        {
          DOMAttributesPtr attrs = pMessageFieldNode->getAttributes();
          std::string required;
          attrs->get("required", required);
          bool isRequired = (required == "Y" || required == "y");
          addXMLComponentFields(pDoc.get(), pMessageFieldNode.get(),
                                msgtype, *this, isRequired);
        }
        if(pMessageFieldNode->getName() == "group")
        {
          DOMAttributesPtr attrs = pMessageFieldNode->getAttributes();
          std::string required;
          attrs->get("required", required);
          bool isRequired = (required == "Y" || required == "y");
          addXMLGroup(pDoc.get(), pMessageFieldNode.get(), msgtype, *this, isRequired);
        }
        RESET_AUTO_PTR(pMessageFieldNode,
                       pMessageFieldNode->getNextSiblingNode());
      }
``

All three of those sections have "or" conditionals where they check for the "group" node.  The group node is always checked for name before even reaching the addXMLGroup private method (the change made in this PR).  

This code is indeed complex already so reworking it for 100% unit test line coverage is outside the scope of this PR.  

If the concern is over the restructuring at some point in the future, the newly added unit tests cover this scenario.  There is a new test called TEST(readFromDocument_GroupNodeMissingNameAttribute) which is expecting a ConfigError as well.  The test ensures that even with restructuring we must always throw a ConfigError if the group node is missing a name attribute.

Reply:
I believe there are at least two solutions to this dilemma:

1) make the test fixture a friend of DataDictionary class and test the otherwise unreachable code in private functions directly and
2) replace the unreachable checks with asserts. 

IMO the second solution should be preferable for low-latency code.

Reply:
Adding a textFixture as a friend to the DataDictionary crosses the line of mixing test code into production code.

The second solution would still result in not getting 100% coverage though as the assertion statement would not be hit (unless I am missing something an LCOV skips assertions for coverage metrics).